### DISCUSSION TITLE
about cluster algorithm

### DISCUSSION THREAD

**Discussion by Lilong08** [futurewei-cloud/network_ml#8] [General]
1. Method of Parameter selection

    cluster number(Kmeans, Kmeans++, Kmedoids)

2. Evaluation method of dimensionality reduction results

3. method to deal with sparse matrix

    we find the matrix from the dataset is sparse

4. distance metrics used in the clustering algorithm

    euclidean

    cosine

**Comment by xieus** [comment]
@Lilong08 Regarding the sparse matrix, could you give me one example of the sparse matrix? Trying to see how sparse the matrix is. Thanks.

**Reply by Lilong08** [reply]
Hello, Dr.Xie, we get the traffic matrix with size(1866, 1000), one sample has 1000 features. More than 1800 samples have more than 990 zero-elements in their feature vectors.

**Comment by zzengcs** [comment]
@Lilong08
---Besides the listed paper in other posts, could you please understand the appropriate use case (e.g., data distribution) for the following sklean tools? 
FastICA (https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.FastICA.html#sklearn.decomposition.FastICA), TSNE(https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html), NMF (https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.NMF.html), TruncatedSVD (https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.TruncatedSVD.html#sklearn.decomposition.TruncatedSVD)

----For SVD, it has explained_variance_ratio_.
---- Could you please check these papers, compare and summarize the key take-away messages.
Kumar, V., Chhabra, J.K. and Kumar, D., 2014. Performance evaluation of distance metrics in the clustering algorithms. INFOCOMP Journal of Computer Science, 13(1), pp.38-52.
S. Kapil and M. Chawla, "Performance evaluation of K-means clustering algorithm with various distance metrics," 2016 IEEE 1st International Conference on Power Electronics, Intelligent Control and Energy Systems (ICPEICES), 2016, pp. 1-4, doi: 10.1109/ICPEICES.2016.7853264.
Foss, Alexander H., Marianthi Markatou, and Bonnie Ray. "Distance metrics and clustering methods for mixed‐type data." International Statistical Review 87, no. 1 (2019): 80-109.

**Reply by Plutozyh** [reply]
[Distance_metrics_report_ZYH.pdf](https://github.com/futurewei-cloud/network_ml/files/9300599/Distance_metrics_report_ZYH.pdf)
Hi, Dr Zeng,
     This is the reading report of three papers on distance measurement, of which the third one is less related to our scenario, so I wrote the third report simply. Please check the details in the attachment. Thanks.

**Comment by zzengcs** [comment]
@Lilong08 Could you please also check the related tools in scipy?  https://docs.scipy.org/doc/scipy/reference/sparse.html

**Comment by Lilong08** [comment]
@zzengcs Hi, Dr.Zeng, in section `3.5 Cluster merging` of the paper, author uses a fixed merging threshold $\theta$ to merge clusters, and the method is inspried by [MeanShift](https://ieeexplore.ieee.org/abstract/document/1000236), if the `cosine` distance of two clusters centroids is less than $\theta$, the two clusters will be merged. The paper does not tell about how to determine $\theta$. Could you please give me some advice on this parameter? Thanks.

**Reply by zzengcs** [reply]
Hello, @Lilong08 , the threshold is a cut-off point in the algorithm. In the implementation of meanshift, usually can start at the value 0f 0.5*bandwidth, and gradually increase/decrease it based on the comparison of model output with the ground truth.
